Zend/zend_execute.c

```diff
@@ -776,6 +776,72 @@ static int zend_verify_internal_arg_type(zend_function *zf, uint32_t arg_num, zv
 	return 1;
 }
 
+zend_bool zend_verify_property_type(zend_class_entry *scope, zend_string *name, zend_uchar type, zend_string *type_name, zend_class_entry **type_ce, zval *property, zend_bool runtime) {
+	switch (type) {
+		case IS_OBJECT: {
+			zend_class_entry *pce = type_ce ? *type_ce : NULL;
+			zend_string *resolved = runtime ? zend_resolve_property_type(type_name, scope) : type_name;
+			
+			if (!pce && runtime) {
+				pce = zend_lookup_class(resolved);
+				
+				if (type_ce) {
+					*type_ce = pce;				
+				}
+			}
+
+			if (Z_TYPE_P(property) != IS_OBJECT || !runtime || !instanceof_function(pce, Z_OBJCE_P(property))) {
+				if (runtime) {
+					zend_throw_exception_ex(zend_ce_type_error, type, 
+						"Typed property %s::$%s must be an instance of %s, %s used",
+							ZSTR_VAL(scope->name),
+							ZSTR_VAL(name),
+							ZSTR_VAL(resolved),
+							Z_TYPE_P(property) == IS_OBJECT ?
+								ZSTR_VAL(Z_OBJCE_P(property)->name) :
+								zend_get_type_by_const(Z_TYPE_P(property)));
+				} else {
+					zend_error(scope->type == ZEND_USER_CLASS ? E_COMPILE_ERROR : E_CORE_ERROR,
+						"Typed property %s::$%s must be an instance of %s, %s used",
+							ZSTR_VAL(scope->name),
+							ZSTR_VAL(name),
+							ZSTR_VAL(resolved),
+							zend_get_type_by_const(Z_TYPE_P(property)));
+				}
+				return 0;
+			}
+			return 1;
+		}
+
+		case IS_LONG:
+			if (Z_TYPE_P(property) == IS_DOUBLE)
+				return 1;
+
+		default:
+			if (!ZEND_SAME_FAKE_TYPE(type, Z_TYPE_P(property))) {
```

Comment:
Shouldn't we handle the case where you have nullable types?
I don't see code anywhere, but this is possible assuming we have:

```php
class Foo {
  public Bar bar;
}
```

Foo::$bar should accept null too (it was the default initial state).